gusto/time_discretisation.py

```diff
@@ -159,21 +173,52 @@ def setup(self, equation, apply_bcs=True, *active_labels):
         # -------------------------------------------------------------------- #
 
         if self.wrapper is not None:
-            self.wrapper.setup()
-            self.fs = self.wrapper.function_space
-            if self.solver_parameters is None:
-                self.solver_parameters = self.wrapper.solver_parameters
-            new_test = TestFunction(self.wrapper.test_space)
-            # SUPG has a special wrapper
-            if self.wrapper_name == "supg":
-                new_test = self.wrapper.test
-
-            # Replace the original test function with the one from the wrapper
-            self.residual = self.residual.label_map(
-                all_terms,
-                map_if_true=replace_test_function(new_test))
+            if self.wrapper_name == "mixed_options":
+
+                self.wrapper.wrapper_spaces = equation.spaces
+                self.wrapper.field_names = equation.field_names
+
+                for field, subwrapper in self.wrapper.subwrappers.items():
+
+                    if field not in equation.field_names:
+                        raise ValueError(f"The option defined for {field} is for a field that does not exist in the equation set")
+
+                    field_idx = equation.field_names.index(field)
+
+                    # Store the original space of the tracer
+                    subwrapper.tracer_fs = self.equation.spaces[field_idx]
 
-            self.residual = self.wrapper.label_terms(self.residual)
+                    subwrapper.setup()
+
+                    # Update the function space to that needed by the wrapper
+                    self.wrapper.wrapper_spaces[field_idx] = subwrapper.function_space
+
+                self.wrapper.setup()
+                self.fs = self.wrapper.function_space
+                new_test_mixed = TestFunctions(self.fs)
+
+                # Replace the original test function with one from the new
+                # function space defined by the subwrappers
+                self.residual = self.residual.label_map(
+                    all_terms,
+                    map_if_true=replace_test_function(new_test_mixed))
+
+            else:
+                self.wrapper.setup()
```

Comment:
```suggestion
                self.wrapper.setup(self.fs)
```
I think this should be `self.fs` but I'm not sure...? Maybe it is just `None`

Reply:
I think we would want this to be None, as we want self.original_space to be None when not using the mixed wrapper.